README.md
# msgpack-nodejs

![GitHub Workflow Status](https://img.shields.io/github/actions/workflow/status/artyomliou/msgpack-nodejs/node.js.yml)
![node-current](https://img.shields.io/node/v/msgpack-nodejs)
![Vulnerabilities](https://img.shields.io/snyk/vulnerabilities/github/artyomliou/msgpack-nodejs)
[![Maintainability](https://api.codeclimate.com/v1/badges/d0d52768bb6be694536e/maintainability)](https://codeclimate.com/github/artyomliou/msgpack-nodejs/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/d0d52768bb6be694536e/test_coverage)](https://codeclimate.com/github/artyomliou/msgpack-nodejs/test_coverage)

Yet another javascript/nodejs implementation of [MsgPack Spec](https://github.com/msgpack/msgpack/blob/master/spec.md).  
The purpose behind is learning by doing, which focuses on modern tools/techniques of nodejs/typescript ecosystem.

# Contents

- [Usage](#usage)
  - [Examples](#examples)
  - [API](#api)
  - [Options](#options)
- [Project status](#project-status)
  - [Compability](#compability)
  - [Limitation](#limitation)
  - [TODO](#todo)
  - [Benchmark](#benchmark)
- [Implementation detail](#implementation-detail)
  - [Encode](#encode)
  - [Decode](#decode)
  - [Lesson learned](#lessons-learned)
- [References](#references)

---

# Usage

```bash
npm i msgpack-nodejs
npm test
```

## Example

Please check [example.md](example.md)

## API

1. [encode()](src/encoder/encoder.ts): `any` => `Uint8Array`
2. [decode()](src/decoder/decoder.ts): `Uint8Array` => `Exclude<any, Map>`
3. [EncodeStream class](src/streams/encode-stream.ts): `Exclude<any, null>` => `Buffer`
4. [DecodeStream class](src/streams/decode-stream.ts): `Buffer` => `Exclude<any, null>` [^1]
5. [registerExtension()](src/extensions/registry.ts): Register your own extension
6. [stringBufferStat()](src/encoder/string-buffer.ts): Show string buffer copied count and size
7. [lruCacheStat()](src/encoder/lru-cache.ts): Show cache hit/miss count
8. [bufferAllocatorStat()](src/encoder/byte-array.ts): Show how byte-array allocate new buffer
9. [prefixTrieStat()](src/decoder/prefix-trie.ts): Show prefix-trie hit/miss count
10. [applyOptions()](src/options.ts): Manually control caching

[^1]: After attaching another stream that does not expect a object as its input, you may encounter error

## Options

You can apply options [like this](example.md#apply-options)

| Key                               | type    | default | Description                                                                                                                    |
| --------------------------------- | ------- | :-----: | ------------------------------------------------------------------------------------------------------------------------------ |
| encoder.mapKeyCache.enabled       | boolean |  true   | Cache map-key or not                                                                                                           |
| encoder.mapKeyCache.size          | number  |   30    | How big is the mapKeyCache                                                                                                     |
| encoder.stringCache.enabled       | boolean |  true   | Cache any string except map-key or not                                                                                         |
| encoder.stringCache.size          | number  |   100   | How big is the stringCache                                                                                                     |
| encoder.byteArray.base            | number  |  1024   | How many bytes will be allocated for every execution. Setting this would increase performance when handling many big JSON data |
| decoder.shortStringCache.enabled  | boolean |  true   | Use prefix-trie or not                                                                                                         |
| decoder.shortStringCache.lessThan | number  |   10    | Only cache if string is shorter than this value                                                                                |
| decoder.jsUtf8Decode.enabled      | boolean |  true   | Use JS utf8-decode or not                                                                                                      |
| decoder.jsUtf8Decode.lessThan     | number  |   200   | Only use JS utf8-decode if string is shorter than this value                                                                   |

---

# Project status

## Compability

| Env        | Executable? |
| ---------- | ----------- |
| Node.js 18 | ✅          |
| Node.js 16 | ✅          |
| Node.js 14 | ✅          |
| Node.js 12 | ❌          |

## Limitation

1. Does not support float 32 encoding, because Javascript float is always 64-bit.

## TODO

1. Ext tests
2. Map 16/32 tests

## Benchmark

By utlizing the great [benchmark tool by msgpack-lite](https://github.com/kawanet/msgpack-lite/blob/master/lib/benchmark.js),
I thought the performance of this project would not be disappointing.

Runs on node.js 16 & laptop with R5-5625U.

| operation                                                          |      op |   ms |       op/s |
| ------------------------------------------------------------------ | ------: | ---: | ---------: |
| buf = Buffer(JSON.stringify(obj));                                 | 1021200 | 5000 |     204240 |
| obj = JSON.parse(buf);                                             | 1279500 | 5000 |     255900 |
| buf = require("msgpack-lite").encode(obj);                         |  685800 | 5000 |     137160 |
| obj = require("msgpack-lite").decode(buf);                         |  389800 | 5001 |      77944 |
| buf = Buffer(require("msgpack.codec").msgpack.pack(obj));          |  713600 | 5000 |     142720 |
| obj = require("msgpack.codec").msgpack.unpack(buf);                |  401300 | 5001 |      80243 |
| buf = require("msgpack-js-v5").encode(obj);                        |  284400 | 5000 |      56880 |
| obj = require("msgpack-js-v5").decode(buf);                        |  544600 | 5000 |     108920 |
| buf = require("msgpack-js").encode(obj);                           |  277100 | 5001 |      55408 |
| obj = require("msgpack-js").decode(buf);                           |  559800 | 5000 |     111960 |
| buf = require("msgpack5")().encode(obj);                           |  147700 | 5001 |      29534 |
| obj = require("msgpack5")().decode(buf);                           |  239500 | 5000 |      47900 |
| buf = require("notepack").encode(obj);                             | 1041500 | 5000 |     208300 |
| obj = require("notepack").decode(buf);                             |  671300 | 5000 |     134260 |
| obj = require("msgpack-unpack").decode(buf);                       |  163400 | 5001 |      32673 |
| **buf = require("msgpack-nodejs").encode(obj);** (Run in sequence) | 1148900 | 5000 | **229780** |
| **obj = require("msgpack-nodejs").decode(buf);** (Run in sequence) |  777500 | 5000 | **155500** |
| **buf = require("msgpack-nodejs").encode(obj);** (Run exclusively) | 1321900 | 5000 | **264380** |
| **obj = require("msgpack-nodejs").decode(buf);** (Run exclusively) |  805400 | 5000 | **161080** |

---

# Implementation detail

## Encode

[Encoder](src/encoder/encoder.ts) uses a recursive function `match()` to match JSON structure (primitive value, object, array or nested), and pushes anything encoded into [ByteArray](src/encoder/byte-array.ts) that responsible for allocating buffer. Encoded string will be written in [StringBuffer](src/encoder/string-buffer.ts) first and cached in [LruCache](src/encoder/lru-cache.ts).

## Decode

[Decoder](src/decoder/decoder.ts) uses `parseBuffer()` to read every value out, and push them into [StructBuilder](src/decoder/struct-builder.ts) to rebuild whole JSON object. For string less than 200 bytes, use pure JS [utf8Decode()](src/decoder/utf8-decode.ts), then cache in [prefix trie](src/decoder/prefix-trie.ts).

## Optimization strategies:

##### Cache

- To improve encoding performance, [LruCache](src/encoder/lru-cache.ts) was used for caching encoded string and its header.
- To improve decoding peformance, [prefix trie](src/decoder/prefix-trie.ts) was deployed for Uint8Array caching.
- To avoid evicting, map-key caching and string caching were separated.

##### ArrayBuffer / TypedArray

- To efficiently allocate new buffer, every [ByteArray](src/encoder/byte-array.ts) begins with small buffer (1K). [^2]
- To efficiently handle unpredictable large JSON, [ByteArray](src/encoder/byte-array.ts) allocates exponentially.
- To avoid overhead on writing, [ByteArray](src/encoder/byte-array.ts) uses [`DataView` calls](https://v8.dev/blog/dataview) as much as possible.

##### Node.js

- To maximize performance of array, use [pre-allocated array](src/decoder/decoder.ts#L11-L12). [^3]
- To maximize performance of string encoding, string are encoded in [StringBuffer](src/encoder/string-buffer.ts) with [encodeInto()](https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder/encodeInto) to prevent unnecessary copying. Then these encoded content will be referenced by `subarray()` for writing and caching. [^4]
- To avoid overhead of `TextDecoder()`, [decode UTF-8 bytes with pure JS](src/decoder/utf8-decode.ts) when less than 200 bytes. [^3]
- To avoid GC overhead in decoder, every parsed typed value will be passed into `builder.insertValue()` directly. [^17]
- To avoid syntax penalty of [private class fields](https://v8.dev/blog/faster-class-features) under node.js 18, use [TypeScript's syntax](https://www.typescriptlang.org/docs/handbook/2/classes.html#caveats) instead.

## Lessons learned

- Javascript
  - The difference between `ArrayBuffer`, `TypedArray` and the node.js API `Buffer` [^5]
  - `BigInt` operators [^6]
  - `left shift` and `right shift` [^7] [^8]
  - Private class features [^9]
  - Pre-allocated array [^3]
  - UTF-8 encoding/decoding [^10]
  - Node.js transform stream [^11]
- Node.js
  - Profiler [^12]
- Typescript
  - Testing - ts-jest [^13]
  - Linter - typescript-eslint & lint-staged [^14]
  - Packaging for ESModule & CommonJS [^15]
- CI & CD
  - Github Actions [^16]

---

[^2]: Thanks to [kriszyp/msgpackr](https://github.com/kriszyp/msgpackr/blob/master/pack.js#L636-L657) for inspiration of better buffer allocation strategy.
[^3]: Thanks to [AppSpector](https://appspector.com/blog/how-to-improve-messagepack-javascript-parsing-speed-by-2-6-times), this article gives very practical advices including pre-allocated array and manual decoding under 200 characters.
[^4]: Thanks to [msgpack/msgpack-javascript](https://github.com/msgpack/msgpack-javascript/blob/da998c654fbba8952c49ec407c554cc7400b36ac/src/Encoder.ts#L178-L195) for technique including UTF-8 bytes calculation and usage of `encodeInto()`, which led me to the ultra optimization strategy.
[^5]: [[JS] TypedArray, ArrayBuffer 和 DataView](https://pjchender.dev/javascript/js-typedarray-buffer-dataview/)
[^6]: [BigInt operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#operators)
[^7]: [left shift](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Left_shift)
[^8]: [right shift](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Right_shift)
[^9]: [Private class features](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields)
[^10]: [UTF-8 encoding/decoding](https://zh.wikipedia.org/zh-tw/UTF-8#UTF-8%E7%9A%84%E7%B7%A8%E7%A2%BC%E6%96%B9%E5%BC%8F)
[^11]: [samerbuna/efficient-node](https://github.com/samerbuna/efficient-node/blob/main/400-node-streams.adoc)
[^12]: [Profiler](https://nodejs.org/en/docs/guides/simple-profiling/)
[^13]: [ts-jest](https://kulshekhar.github.io/ts-jest/docs/guides/esm-support)
[^14]: [使用 ESLint, Prettier, Husky, Lint-staged 以及 Commitizen 提升專案品質及一致性](https://medium.com/@danielhu95/set-up-eslint-pipeline-zh-tw-990d7d9eb68e)
[^15]: [Best practices for creating a modern npm package](https://snyk.io/blog/best-practices-create-modern-npm-package/)
[^16]: [Github Actions](https://github.com/artyomliou/msgpack-nodejs/actions)
[^17]: [Memory Diagnostics - Using GC Trace](https://nodejs.org/en/docs/guides/diagnostics/memory/using-gc-traces/)


CHANGELOG.md
# Versions

## 1.3.9

- optimize(decoder): huge improvement by reducing GC
- fix: code of `typed-value-resolver.ts` was merged into `decoder.ts`
- style: update var naming & add comments

## 1.3.8

- optimize(encoder): encode string in specific buffer
- optimize(encoder): cache header and encoded at once
- optimize(encoder): disabled noRareKeys() because of its cost

## 1.3.7

- feat: more available options
- optimize(encoder): different size for string cache
- test: rewrite encoder/decoder tests

## 1.3.6

- optimize: use typescript's private class field. [^1]
- optimize: lower cost of passing cb into LruCache remember()
- optimize(encoder): exponential buffer allocation
- optimize(encoder): separated cache of map-key & string
- optimize(decoder): cache short string(<10) in tree [^2]
- feat: applyOptions()
- fix: remove utf8 encoding implementation

Again, encoding and decoding [aws-ip-ranges.json](test/dataset/aws-ip-ranges.json).

encoding: 40ms -> 35ms

[^1]: [The performance of private class fields was bad before Node.js 18](https://v8.dev/blog/faster-class-features), and [private and protected in Typescript are only are only enforced during type checking](https://www.typescriptlang.org/docs/handbook/2/classes.html#caveats), so it's possible that this change would bring performance improvement.<br>
[^2]: [Inserting numeric integer key in object was pretty fast](https://github.com/artyomliou/benchmark-js-object-addition), and uint8 is great candidate. To avoid too many random access bring negative optimization, just cache short string.

## 1.3.5

- refactor(decoder): for better readability & performance
- feat(decoder): implemented UTF-8 decoding for optimization
- optimize(encoder): better and simpler allocating strategy
- optimize: encoder/decoder perf

Again, encoding and decoding [aws-ip-ranges.json](test/dataset/aws-ip-ranges.json).

encoding: 50ms -> 40ms
decoding: 100ms -> 45ms

## 1.3.4

- optimize: encoder/decoder perf

Again, encoding and decoding [aws-ip-ranges.json](test/dataset/aws-ip-ranges.json).

encoding: 50ms
decoding: 150ms -> 100ms

## 1.3.3

- optimize: encoder/decoder perf with caching and profiler

Again, encoding and decoding [aws-ip-ranges.json](test/dataset/aws-ip-ranges.json).

encoding: 60ms -> 50ms
decoding: 250ms -> 150ms

## 1.3.2

- fix: ESM module import
- fix: **truly improve encoder performance** with larger page & caching map

For example, encoding [aws-ip-ranges.json](test/dataset/aws-ip-ranges.json) in previous release takes 1 second.
Now it takes only 60ms, which improve by 94%.

## 1.3.1

- optimize: improve encoder performance by caching on map and its string key

## 1.3.0

- feat: support custom extension
- BREAKING CHANGE: use Uint8Arry for most encoding/decoding

## 1.2.0

- **refactor: migration to Typescript**
- style: use **typescript-eslint** instead
- style: **prettify on save** (through [.vscode/settings.json](.vscode/settings.json))
- style: use **lint-staged**
- style: use **kebab case** for .ts files
- test: use **ts-jest** instead because of integrating Mocha with typescript is difficult
- build: temporarily suspend support of browser
- build: support both **ES module** & **CommonJS module**

## 1.1.0

- ci: Create node.js.yml
- feat: supports stream
- fix: re-structure this project
- fix(docs): updated with runnable code
- feat(build): Separate entries of browser & server

## 1.0.1

- Support browser with esbuild
- Minify dependency size with .npmignore


src/Serialize.js
import ByteArray from './ByteArray.js'
import {
  NIL,
  BOOL_FALSE,
  BOOL_TRUE,
  BIN8_PREFIX,
  BIN16_PREFIX,
  BIN32_PREFIX,
  EXT8_PREFIX,
  EXT16_PREFIX,
  EXT32_PREFIX,
  // FLOAT32_PREFIX,
  FLOAT64_PREFIX,
  UINT8_PREFIX,
  UINT16_PREFIX,
  UINT32_PREFIX,
  UINT64_PREFIX,
  INT8_PREFIX,
  INT16_PREFIX,
  INT32_PREFIX,
  INT64_PREFIX,
  FIXEXT1_PREFIX,
  FIXEXT2_PREFIX,
  FIXEXT4_PREFIX,
  FIXEXT8_PREFIX,
  FIXEXT16_PREFIX,
  STR8_PREFIX,
  STR16_PREFIX,
  STR32_PREFIX,
  ARRAY16_PREFIX,
  ARRAY32_PREFIX,
  MAP16_PREFIX,
  MAP32_PREFIX,
  EXT_TYPE_TIMESTAMP
} from './constants/index.js'
import TimeSpec from './TimeSpec.js'

/**
 * @param {Object} srcObject
 * @param {boolean} debug
 */
export default function messagePackSerialize (srcObject, debug = false) {
  const byteArray = new ByteArray()
  match(byteArray, srcObject)

  const buffer = byteArray.getBuffer()
  if (debug) {
    console.debug(buffer)
  }
  return buffer
}

/**
 * @param {ByteArray} byteArray
 * @param {*} val
 * @returns
 */
function match (byteArray, val) {
  switch (typeof val) {
    case 'boolean':
      handleBoolean(byteArray, val)
      break

    case 'bigint':
      handleInteger(byteArray, val)
      break

    case 'number':
      if (Number.isInteger(val)) {
        handleInteger(byteArray, val)
      } else {
        handleFloat(byteArray, val)
      }
      break

    case 'string':
      handleString(byteArray, val)
      break

    case 'object':
      if (val === null) {
        byteArray.writeUint8(NIL)
        break
      }
      if (val instanceof Date) {
        handleTimestamp(byteArray, val)
        break
      }
      if (val instanceof ArrayBuffer) {
        handleBuffer(byteArray, val)
        break
      }
      if (Array.isArray(val)) {
        handleArray(byteArray, val)
        for (const element of val) {
          match(byteArray, element)
        }
        break
      }

      // Handling typical object
      handleMap(byteArray, val)
      for (const [k, v] of (val instanceof Map ? val.entries() : Object.entries(val))) {
        handleString(byteArray, k)
        match(byteArray, v)
      }
      break

    default:
      console.debug('noop', val)
      // No support for Symbol, Function, Undefined
      break
  }
}

/**
 * @param {ByteArray} byteArray
 * @param {Number} number
 * @returns
 */
function handleBoolean (byteArray, val = true) {
  if (val) {
    byteArray.writeUint8(BOOL_TRUE)
  } else {
    byteArray.writeUint8(BOOL_FALSE)
  }
}

/**
 * @param {ByteArray} byteArray
 * @param {Number} number
 * @returns
 */
function handleInteger (byteArray, number = 0) {
  // positive fixint stores 7-bit positive integer
  // 0b1111111 = 127
  if (number >= 0 && number <= 127) {
    byteArray.writeUint8(number)
    return
  }

  // negative fixint stores 5-bit negative integer
  if (number < 0 && number >= -32) {
    byteArray.writeInt8(number)
    return
  }

  // unsigned
  if (number > 0) {
    if (number <= 0xFF) {
      byteArray.writeUint8(UINT8_PREFIX)
      byteArray.writeUint8(number)
      return
    }
    if (number <= 0xFFFF) {
      byteArray.writeUint8(UINT16_PREFIX)
      byteArray.writeUint16(number)
      return
    }
    if (number <= 0xFFFFFFFF) {
      byteArray.writeUint8(UINT32_PREFIX)
      byteArray.writeUint32(number)
      return
    }
    byteArray.writeUint8(UINT64_PREFIX)
    byteArray.writeUint64(number)
    return
  }

  // signed
  if (number < 0) {
    if (-number <= 0xFF) {
      byteArray.writeUint8(INT8_PREFIX)
      byteArray.writeInt8(number)
      return
    }
    if (-number <= 0xFFFF) {
      byteArray.writeUint8(INT16_PREFIX)
      byteArray.writeInt16(number)
      return
    }
    if (-number <= 0xFFFFFFFF) {
      byteArray.writeUint8(INT32_PREFIX)
      byteArray.writeInt32(number)
      return
    }
    byteArray.writeUint8(INT64_PREFIX)
    byteArray.writeInt64(number)
  }
}

/**
 * @param {ByteArray} byteArray
 * @param {Number} number
 * @returns
 */
function handleFloat (byteArray, number = 0) {
  // Since all float in Javascript is double, it's not possible to have FLOAT32 type.
  byteArray.writeUint8(FLOAT64_PREFIX)
  byteArray.writeFloat64(number)
}

/**
 * @param {ByteArray} byteArray
 * @param {String} string
 * @returns
 */
function handleString (byteArray, string = '') {
  const strBuf = new TextEncoder().encode(string)
  const bytesCount = strBuf.byteLength
  if (bytesCount <= 31) {
    byteArray.writeUint8(0b10100000 + bytesCount)
    byteArray.writeBuffer(strBuf)
    return
  }
  // (2 ** 8) - 1
  if (bytesCount < 0xFF) {
    byteArray.writeUint8(STR8_PREFIX)
    byteArray.writeUint8(bytesCount)
    byteArray.writeBuffer(strBuf)
    return
  }
  // (2 ** 16) - 1
  if (bytesCount < 0xFFFF) {
    byteArray.writeUint8(STR16_PREFIX)
    byteArray.writeUint16(bytesCount)
    byteArray.writeBuffer(strBuf)
    return
  }
  // (2 ** 32) - 1
  if (bytesCount < 0xFFFFFFFF) {
    byteArray.writeUint8(STR32_PREFIX)
    byteArray.writeUint32(bytesCount)
    byteArray.writeBuffer(strBuf)
    return
  }
  throw new Error('Length of string value cannot exceed (2^32)-1.')
}

/**
 * @param {ByteArray} byteArray
 * @param {Buffer} buffer
 * @returns
 */
function handleBuffer (byteArray, buffer) {
  const bytesCount = buffer.byteLength
  // (2 ** 8) - 1
  if (bytesCount < 0xFF) {
    byteArray.writeUint8(BIN8_PREFIX)
    byteArray.writeUint8(bytesCount)
    byteArray.writeBuffer(buffer)
    return
  }
  // (2 ** 16) - 1
  if (bytesCount < 0xFFFF) {
    byteArray.writeUint8(BIN16_PREFIX)
    byteArray.writeUint16(bytesCount)
    byteArray.writeBuffer(buffer)
    return
  }
  // (2 ** 32) - 1
  if (bytesCount < 0xFFFFFFFF) {
    byteArray.writeUint8(BIN32_PREFIX)
    byteArray.writeUint32(bytesCount)
    byteArray.writeBuffer(buffer)
    return
  }
  throw new Error('Length of binary value cannot exceed (2^32)-1.')
}

/**
 * @param {ByteArray} byteArray
 * @param {Array} array
 * @returns
 */
function handleArray (byteArray, array = {}) {
  const arraySize = array.length

  // fixarray
  if (arraySize < 0xF) {
    byteArray.writeUint8(0b10010000 + arraySize)
    return
  }

  // map 16
  if (arraySize < 0xFFFF) {
    byteArray.writeUint8(ARRAY16_PREFIX)
    byteArray.writeUint16(arraySize)
    return
  }

  // map 32
  if (arraySize < 0xFFFFFFFF) {
    byteArray.writeUint8(ARRAY32_PREFIX)
    byteArray.writeUint32(arraySize)
    return
  }

  throw new Error('Number of elements cannot exceed (2^32)-1.')
}

/**
 * @param {ByteArray} byteArray
 * @param {Object|Map} map
 * @returns
 */
function handleMap (byteArray, map = {}) {
  const mapSize = map instanceof Map ? map.size : Object.keys(map).length

  // fixmap
  if (mapSize < 0xF) {
    byteArray.writeUint8(0b10000000 + mapSize)
    return
  }

  // map 16
  if (mapSize < 0xFFFF) {
    byteArray.writeUint8(MAP16_PREFIX)
    byteArray.writeUint16(mapSize)
    return
  }

  // map 32
  if (mapSize < 0xFFFFFFFF) {
    byteArray.writeUint8(MAP32_PREFIX)
    byteArray.writeUint32(mapSize)
    return
  }

  throw new Error('Number of pairs cannot exceed (2^32)-1.')
}

/**
 * @param {ByteArray} byteArray
 * @param {Date} date
 * @returns
 * @ref https://github.com/msgpack/msgpack/blob/master/spec.md#timestamp-extension-type
 */
function handleTimestamp (byteArray, date) {
  const time = TimeSpec.fromDate(date)
  if (time.nsec > 1000000000) {
    throw new Error('Nanoseconds cannot be larger than 999999999.')
  }

  if (time.sec >= 0 && time.sec <= 0xFFFFFFFF) {
    // (data64 & 0xffffffff00000000L == 0)
    // It is basically doing masking on the data64 variable, which only keep nsec, and decide if it equals to 0.
    if (time.nsec === 0) {
      // timestamp 32
      const view = new DataView(new ArrayBuffer(4))
      view.setUint32(0, Number(time.sec), false) // unsigned
      handleExt(byteArray, EXT_TYPE_TIMESTAMP, view.buffer)
    } else {
      // timestamp 64
      const data64 = (BigInt(time.nsec) << 34n) + BigInt(time.sec)
      const view = new DataView(new ArrayBuffer(8))
      view.setBigUint64(0, data64, false) // unsigned
      handleExt(byteArray, EXT_TYPE_TIMESTAMP, view.buffer)
    }
  } else {
    // timestamp 96
    const view = new DataView(new ArrayBuffer(12))
    view.setUint32(0, time.nsec, false) // unsigned
    view.setBigInt64(4, BigInt(time.sec), false)// signed
    handleExt(byteArray, EXT_TYPE_TIMESTAMP, view.buffer)
  }
}

/**
 * @param {ByteArray} byteArray
 * @param {Number} type
 * @param {ArrayBuffer} data
 * @returns
 */
function handleExt (byteArray, type, data) {
  const byteLength = data.byteLength

  // fixint
  if (byteLength === 1) {
    byteArray.writeUint8(FIXEXT1_PREFIX)
    byteArray.writeInt8(type)
    byteArray.writeBuffer(data)
    return
  } else if (byteLength === 2) {
    byteArray.writeUint8(FIXEXT2_PREFIX)
    byteArray.writeInt8(type)
    byteArray.writeBuffer(data)
    return
  } else if (byteLength === 4) {
    byteArray.writeUint8(FIXEXT4_PREFIX)
    byteArray.writeInt8(type)
    byteArray.writeBuffer(data)
    return
  } else if (byteLength === 8) {
    byteArray.writeUint8(FIXEXT8_PREFIX)
    byteArray.writeInt8(type)
    byteArray.writeBuffer(data)
    return
  } else if (byteLength === 16) {
    byteArray.writeUint8(FIXEXT16_PREFIX)
    byteArray.writeInt8(type)
    byteArray.writeBuffer(data)
    return
  }

  // ext 8
  if (byteLength < 0xFF) {
    byteArray.writeUint8(EXT8_PREFIX)
    byteArray.writeUint8(byteLength)
    byteArray.writeInt8(type)
    byteArray.writeBuffer(data)
    return
  } else if (byteLength < 0xFFFF) {
    byteArray.writeUint8(EXT16_PREFIX)
    byteArray.writeUint16(byteLength)
    byteArray.writeInt8(type)
    byteArray.writeBuffer(data)
    return
  } else if (byteLength < 0xFFFFFFFF) {
    byteArray.writeUint8(EXT32_PREFIX)
    byteArray.writeUint32(byteLength)
    byteArray.writeInt8(type)
    byteArray.writeBuffer(data)
    return
  }

  throw new Error('Ext does not support data exceeding 2**32-1 bytes.')
}
